src/EventStore.Core/Services/Archiver/Storage/FileSystemArchiveStorage.cs

```diff
@@ -16,11 +21,74 @@ public FileSystemArchiveStorage(FileSystemOptions options) {
 		_archivePath = options.Path;
 	}
 
-	public ValueTask StoreChunk(string pathToSourceChunk) {
-		var fileName = Path.GetFileName(pathToSourceChunk);
-		var pathToDestnationChunk = Path.Combine(_archivePath, fileName);
-		Log.Information("Copying file {Path} to {Destination}", pathToSourceChunk, pathToDestnationChunk);
-		File.Copy(pathToSourceChunk, pathToDestnationChunk);
-		return ValueTask.CompletedTask;
+	public async ValueTask<bool> StoreChunk(string chunkPath, CancellationToken ct) {
+		try {
+			var destinationPath = Path.Combine(_archivePath, Path.GetFileName(chunkPath));
+			var tempPath = $"{destinationPath}.tmp";
+
+			if (File.Exists(destinationPath))
+				File.Delete(destinationPath);
+
+			if (File.Exists(tempPath))
+				File.Delete(tempPath);
+
+			await using var source = File.Open(
+				path: chunkPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.Open,
+					Access = FileAccess.Read,
+					Share = FileShare.Read,
+					Options = FileOptions.SequentialScan | FileOptions.Asynchronous
+				});
+
+			await using var destination = File.Open(
+				path: tempPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.CreateNew,
+					Access = FileAccess.ReadWrite,
+					Share = FileShare.None,
+					Options = FileOptions.Asynchronous,
+					PreallocationSize = new FileInfo(chunkPath).Length
+				});
+
+			await source.CopyToAsync(destination, ct);
+
+			File.Move(tempPath, destinationPath);
+
+			return true;
+		} catch (Exception ex) when (ex is not OperationCanceledException) {
+			if (!File.Exists(chunkPath))
+				throw new ChunkDeletedException();
+
+			Log.Error(ex, "Error while storing chunk: {chunkFile}", Path.GetFileName(chunkPath));
+			return false;
+		}
+	}
+
+	public ValueTask<bool> RemoveChunks(int chunkStartNumber, int chunkEndNumber, string exceptChunk, CancellationToken ct) {
+		try {
+			var directoryInfo = new DirectoryInfo(_archivePath);
+			for (var chunkNumber = chunkStartNumber; chunkNumber <= chunkEndNumber; chunkNumber++) {
+				foreach (var file in directoryInfo.EnumerateFiles($"chunk-{chunkNumber:000000}.*")) {
+					if (file.Name == exceptChunk)
+						continue;
+
+					File.Delete(file.FullName);
+				}
+			}
+		} catch (Exception ex) when (ex is not OperationCanceledException) {
```

Comment:
If `ex` is not `OperationCanceledException` then the method throws instead of returning the task with wrapped exception.

Reply:
isn't it the other way round? (if `ex` is `OperationCanceledException`, then it'll throw it, otherwise it'll log an error and return false).

maybe it would be more easily readable if we add an explicit catch for OperationCanceledException:
```
catch (OperationCanceledException) {
   throw;
} catch (Exception ex) {
...
}
```

Reply:
fixed